File: sdk/android/src/main/java/com/vgdp/sdk/VGDPClient.kt
package com.vgdp.sdk

import android.app.Activity
import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

enum class VGDPEnvironment {
    AMOY,
    POLYGON
}

data class VGDPConfig(
    val apiBaseUrl: String,
    val environment: VGDPEnvironment,
    val wasmAsset: String,
    val zkeyAsset: String
)

// ---------------------------------------------------------------------------
// Data models
// ---------------------------------------------------------------------------

data class VGDPDeliveryOrder(
    val orderId: String,
    val companyId: String,
    val targetLatE7: Int,
    val targetLonE7: Int,
    val radiusMeters: Int,
    val riderDid: String,
    val riderWallet: String
)

data class VGDPTrackingOptions(
    val activationRadiusMeters: Double = 200.0,
    val pollingIntervalSeconds: Long = 5L,
    val desiredAccuracyMeters: Double = 10.0
)

data class VGDPProofResult(
    val proofId: String,
    val transactionHash: String,
    val blockNumber: Int?,
    val merkleRoot: String,
    val status: String
)

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

sealed class VGDPException(message: String) : Exception(message) {
    class NotImplemented(feature: String) : VGDPException("Not implemented: $feature")
    class OrderNotFound(orderId: String) : VGDPException("No active tracking for order: $orderId")
    class ProofGenerationFailed(reason: String) : VGDPException("Proof generation failed: $reason")
    class ApiError(val status: Int, body: String) : VGDPException("API error $status: $body")
    class LocationUnavailable : VGDPException("Location unavailable")
    class ConfigurationError(reason: String) : VGDPException("Configuration error: $reason")
}

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

class VGDPClient(
    private val context: Context,
    private val config: VGDPConfig
) {
    private val activeOrders = mutableMapOf<String, VGDPDeliveryOrder>()

    suspend fun configureRiderIdentity(did: String, keyAlias: String) {
        TODO("Not implemented: bind rider DID to Android Keystore alias '$keyAlias'")
    }

    suspend fun startTracking(
        order: VGDPDeliveryOrder,
        options: VGDPTrackingOptions = VGDPTrackingOptions()
    ) {
        activeOrders[order.orderId] = order
        TODO("Not implemented: start Fused Location Provider polling every ${options.pollingIntervalSeconds}s")
    }

    suspend fun confirmDelivered(
        orderId: String,
        riderJwt: String,
        activity: Activity
    ): VGDPProofResult {
        val order = activeOrders[orderId]
            ?: throw VGDPException.OrderNotFound(orderId)

        val bundle = buildProofBundle(order, activity)
        return submitBundle(bundle, riderJwt)
    }

    fun stopTracking(orderId: String) {
        activeOrders.remove(orderId)
    }

    // -----------------------------------------------------------------------
    // Private internals — all stubbed
    // -----------------------------------------------------------------------

    private suspend fun buildProofBundle(order: VGDPDeliveryOrder, activity: Activity): String {
        val location = captureLocation()
        val ntpTime = syncNtp()
        val imageData = capturePhoto(activity)
        val pHash = computePhotoPHash(imageData)
        val salt = secureRandomHex32()
        val nonce = secureRandomHex32()

        val proof = generateZKProof(
            order = order,
            actualLatE7 = location.first,
            actualLonE7 = location.second,
            deliveredAtEpoch = ntpTime
        )

        val digest = buildBundleDigest(
            order = order,
            zkProofHash = proof.zkProofHash,
            photoPHash = pHash,
            salt = salt,
            deliveredAtEpoch = ntpTime,
            nonce = nonce
        )

        val signature = signDigest(digest)
        val attestationJwt = playIntegrityAttest(nonce)

        return buildJsonBundle(
            order = order,
            proof = proof,
            pHash = pHash,
            salt = salt,
            nonce = nonce,
            deliveredAtEpoch = ntpTime,
            didSignature = signature,
            attestationJwt = attestationJwt
        )
    }

    private suspend fun captureLocation(): Pair<Int, Int> {
        TODO("Not implemented: capture high-accuracy GPS via Fused Location Provider")
    }

    private suspend fun syncNtp(): Long {
        TODO("Not implemented: NTP-corrected epoch seconds")
    }

    private suspend fun capturePhoto(activity: Activity): ByteArray {
        TODO("Not implemented: launch CameraX intent and await result")
    }

    private fun computePhotoPHash(imageData: ByteArray): String {
        TODO("Not implemented: DCT-based perceptual hash → hex bytes32")
    }

    private fun secureRandomHex32(): String {
        TODO("Not implemented: 32 cryptographically random bytes → 0x-prefixed hex")
    }

    private data class ZKProofOutput(
        val zkProofHash: String,
        val snarkProof: Map<String, Any>,
        val publicSignals: List<String>,
        val solidityProof: Map<String, Any>
    )

    private suspend fun generateZKProof(
        order: VGDPDeliveryOrder,
        actualLatE7: Int,
        actualLonE7: Int,
        deliveredAtEpoch: Long
    ): ZKProofOutput {
        TODO("Not implemented: load .wasm and .zkey from assets, run snarkjs Groth16 prover in background thread")
    }

    private fun buildBundleDigest(
        order: VGDPDeliveryOrder,
        zkProofHash: String,
        photoPHash: String,
        salt: String,
        deliveredAtEpoch: Long,
        nonce: String
    ): ByteArray {
        TODO("Not implemented: keccak256 ABI-encode per VGDP_PROOF_BUNDLE_V1 spec")
    }

    private fun signDigest(digest: ByteArray): String {
        TODO("Not implemented: sign digest with Android Keystore EC key → 0x-prefixed hex signature")
    }

    private suspend fun playIntegrityAttest(nonce: String): String {
        TODO("Not implemented: Play Integrity API nonce-bound attestation JWT")
    }

    private fun buildJsonBundle(
        order: VGDPDeliveryOrder,
        proof: ZKProofOutput,
        pHash: String,
        salt: String,
        nonce: String,
        deliveredAtEpoch: Long,
        didSignature: String,
        attestationJwt: String
    ): String {
        TODO("Not implemented: assemble ProofBundle JSON matching API schema")
    }

    // -----------------------------------------------------------------------
    // HTTP submit — uses URLConnection (no external deps required)
    // -----------------------------------------------------------------------

    private suspend fun submitBundle(bundleJson: String, riderJwt: String): VGDPProofResult =
        withContext(Dispatchers.IO) {
            val url = URL("${config.apiBaseUrl}/proofs")
            val conn = url.openConnection() as HttpURLConnection
            try {
                conn.requestMethod = "POST"
                conn.setRequestProperty("Content-Type", "application/json")
                conn.setRequestProperty("Authorization", "Bearer $riderJwt")
                conn.doOutput = true

                OutputStreamWriter(conn.outputStream, Charsets.UTF_8).use { it.write(bundleJson) }

                if (conn.responseCode !in 200..299) {
                    val body = conn.errorStream?.bufferedReader()?.readText() ?: ""
                    throw VGDPException.ApiError(conn.responseCode, body)
                }

                val body = conn.inputStream.bufferedReader().readText()
                val json = JSONObject(body)

                VGDPProofResult(
                    proofId = json.getString("proofId"),
                    transactionHash = json.getString("transactionHash"),
                    blockNumber = if (json.has("blockNumber")) json.getInt("blockNumber") else null,
                    merkleRoot = json.getString("merkleRoot"),
                    status = json.getString("status")
                )
            } finally {
                conn.disconnect()
            }
        }
}

File: sdk/ios/Sources/VGDP/VGDPClient.swift
import Foundation

// MARK: - Configuration

public struct VGDPConfig {
    public let apiBaseURL: URL
    public let environment: VGDPEnvironment
    public let zkArtifacts: VGDPZKArtifacts

    public init(apiBaseURL: URL, environment: VGDPEnvironment, zkArtifacts: VGDPZKArtifacts) {
        self.apiBaseURL = apiBaseURL
        self.environment = environment
        self.zkArtifacts = zkArtifacts
    }
}

public enum VGDPEnvironment {
    case amoy
    case polygon
}

public struct VGDPZKArtifacts {
    public let wasmName: String
    public let zkeyName: String
    public let verificationKeyName: String

    public static func bundled(
        wasmName: String,
        zkeyName: String,
        verificationKeyName: String
    ) -> VGDPZKArtifacts {
        VGDPZKArtifacts(
            wasmName: wasmName,
            zkeyName: zkeyName,
            verificationKeyName: verificationKeyName
        )
    }
}

// MARK: - Data Models

public struct VGDPDeliveryOrder {
    public let orderId: String
    public let companyId: String
    public let targetLatE7: Int32
    public let targetLonE7: Int32
    public let radiusMeters: UInt32
    public let riderDid: String
    public let riderWallet: String

    public init(
        orderId: String,
        companyId: String,
        targetLatE7: Int32,
        targetLonE7: Int32,
        radiusMeters: UInt32,
        riderDid: String,
        riderWallet: String
    ) {
        self.orderId = orderId
        self.companyId = companyId
        self.targetLatE7 = targetLatE7
        self.targetLonE7 = targetLonE7
        self.radiusMeters = radiusMeters
        self.riderDid = riderDid
        self.riderWallet = riderWallet
    }
}

public struct VGDPTrackingOptions {
    public let activationRadiusMeters: Double
    public let pollingIntervalSeconds: TimeInterval
    public let desiredAccuracyMeters: Double

    public init(
        activationRadiusMeters: Double = 200,
        pollingIntervalSeconds: TimeInterval = 5,
        desiredAccuracyMeters: Double = 10
    ) {
        self.activationRadiusMeters = activationRadiusMeters
        self.pollingIntervalSeconds = pollingIntervalSeconds
        self.desiredAccuracyMeters = desiredAccuracyMeters
    }
}

public struct VGDPProofResult {
    public let proofId: String
    public let transactionHash: String
    public let blockNumber: Int?
    public let merkleRoot: String
    public let status: String
}

// MARK: - Errors

public enum VGDPError: Error {
    case notImplemented
    case orderNotFound(String)
    case proofGenerationFailed(String)
    case apiError(Int, String)
    case locationUnavailable
    case configurationError(String)
}

// MARK: - Client

public final class VGDPClient {
    private let config: VGDPConfig
    private var activeOrders: [String: VGDPDeliveryOrder] = [:]
    private let session = URLSession.shared

    public init(config: VGDPConfig) {
        self.config = config
    }

    public func configureRiderIdentity(did: String, keyAlias: String) async throws {
        throw VGDPError.notImplemented
    }

    public func startTracking(
        order: VGDPDeliveryOrder,
        options: VGDPTrackingOptions = VGDPTrackingOptions()
    ) async throws {
        activeOrders[order.orderId] = order
        throw VGDPError.notImplemented
    }

    public func confirmDelivered(
        orderId: String,
        riderJWT: String,
        requirePhoto: Bool = true
    ) async throws -> VGDPProofResult {
        guard let order = activeOrders[orderId] else {
            throw VGDPError.orderNotFound(orderId)
        }
        let bundle = try await _buildProofBundle(order: order, riderJWT: riderJWT)
        return try await _submitBundle(bundle: bundle, riderJWT: riderJWT)
    }

    public func stopTracking(orderId: String) {
        activeOrders.removeValue(forKey: orderId)
    }

    // MARK: - Private

    private func _buildProofBundle(order: VGDPDeliveryOrder, riderJWT: String) async throws -> Data {
        throw VGDPError.notImplemented
    }

    private func _captureLocation() async throws -> (latE7: Int32, lonE7: Int32) {
        throw VGDPError.notImplemented
    }

    private func _capturePhoto() async throws -> Data {
        throw VGDPError.notImplemented
    }

    private func _computePhotoPHash(_ imageData: Data) throws -> Data {
        throw VGDPError.notImplemented
    }

    private func _generateZKProof(input: [String: String]) async throws -> Data {
        throw VGDPError.notImplemented
    }

    private func _signDigest(_ digest: Data) throws -> Data {
        throw VGDPError.notImplemented
    }

    private func _submitBundle(bundle: Data, riderJWT: String) async throws -> VGDPProofResult {
        var request = URLRequest(url: config.apiBaseURL.appendingPathComponent("proofs"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(riderJWT)", forHTTPHeaderField: "Authorization")
        request.httpBody = bundle

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw VGDPError.apiError(0, "Invalid response")
        }
        guard (200...299).contains(httpResponse.statusCode) else {
            throw VGDPError.apiError(httpResponse.statusCode, String(data: data, encoding: .utf8) ?? "")
        }

        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        return VGDPProofResult(
            proofId: json?["proofId"] as? String ?? "",
            transactionHash: json?["transactionHash"] as? String ?? "",
            blockNumber: json?["blockNumber"] as? Int,
            merkleRoot: json?["merkleRoot"] as? String ?? "",
            status: json?["status"] as? String ?? ""
        )
    }
}
